---
title: "Notes"
author: "Jinliang Yang"
date: "July 23, 2015"
output: pdf_document
---

# GBS data Summary

## Missing rate and MAF of GBS data

Loading HDF5 format GBS raw data
 - loading in genotypes from HDF5 file `largedata/teo.h5`
 - filtering biallelic loci: Removed `357,647` non-biallelic loci.
 - data matrix dimension: [1:598043, 1:4875]

The missing rates were plotted as below for 598,043 SNPs of 4,875 plants (70/4,875 (`r round(70/4875*100, 1)`%) are founder lines). Note several plants have very high SNP missing rate, i.e. > 80%. Some of them even have a 100% missing rate. In addition, ~20% of them have very high (>80%) per locus missing rate.

```{r}
# source("../profiling/2.ci_data/2.A.1_vsb_hdf5.R")
info <- read.csv("../data/teo_info.csv")
imiss <- read.csv("../data/teo_imiss.csv")

par(mfrow=c(1,2))
hist(imiss$imiss, main="per plant (N=4,875)", col="#008080", breaks=100, xlab="missing rate")
abline(v=0.8, col="red", lty=2, lwd=3)
hist(info$lmiss, main="per locus (N=598,043)", col="#008080", breaks=100, xlab="missing rate")
abline(v=0.8, col="red", lty=2, lwd=3)
```


```{r, eval=TRUE}
par(mfrow=c(1,2))
hist(info$maf, main="Minor Allele Freq (N=598,043)", col="#008080", breaks=100, xlab="MAF")
abline(v=0.8, col="red", lty=2, lwd=3)

hist(subset(info, maf>0.01)$maf, main="MAF >0.01 (N=401,352)", col="#008080", breaks=100, xlab="MAF")
abline(v=0.8, col="red", lty=2, lwd=3)
```

See above the MAF plots. Note `r nrow(subset(info, maf<0.01))` SNPs have the MAF < 0.01.


# Parentage Infomation

```{r}
# source("../profiling/2.ci_data/2.B.1_cj_parentage.R")
pinfo <- read.table("../data/parentage_sum.txt", header=TRUE)
#dim(pinfo) 68,5
subset(pinfo, !is.na(WGS))
subset(pinfo, nox < 30)

par(mfrow=c(1,1))
counts <- pinfo[, c("nselfer", "nox")]
counts[is.na(counts)] <- 0
counts$tot <- counts$nselfer + counts$nox
counts <- counts[order(counts$tot, decreasing=TRUE), ]
counts <- as.matrix(t(counts[,1:2]))
barplot(counts, main="Size Distribution", xlab="families", col=c("darkblue","red"), 
        legend = c("size of selfing", "size of outcrossing"))

```

We calculated the number of selfers and the number of outcrossers associated with the founder lines. In total, `r nrow(pinfo)` unique founder lines were involved in the crosses. Among them, `r nrow(subset(pinfo, !is.na(nselfer)))` founder lines had selfing families; and the bottom ten smallest family had their family size of `r head(sort(pinfo$nselfer[!is.na(pinfo$nselfer)]), 10)`, respectively.   


# WGS of 19 _Teosintes_

```{r, eval=FALSE}
# source("../profiling/2.ci_data/2.B.2_cj_wgsdata.R")

source("../profiling/2.ci_data/2.B.3_cj_comp.R")

### load WGS data of 19 teosintes and recoded to `0, 1, 2` format and `3` indicates missing.
wgs <- recode()
###>>> WGS [ 396818 ] | GBS [ 597607 ] | shared [ 315514 ]
###>>> consistent SNP calling [ 301249 ]

### load GBS data of 19 teosintes
gbs <- gbsgeno(wgs)
###>>> GBS of [ 598043 ] SNPs and [ 19 ] plants
###>>> Common SNPs [ 301249 ] 

### estimate the GBS SNP calling error rate
res <- comp_alleles(wgs, gbs)
###>>> Heterozygote error rate [ 49.1 ] and Homozygote error rate [ 1.7 ]
###>>> het err=[ 494582 ]; het tot=[ 1008196 ]; hom err=[ 45239 ]; hom err=[ 2714395 ]

### calculate missing rate and MAF for 19 teosintes
maf_missing(wgs, gbs)
###>>> Data write to: [ cache/teo_gbs_wgs.RData]

```


From this file `/group/jrigrp4/phasing/cj_teosinte/genotypes_teosinte_19_noScaffolds_or_organelles.geno`, a total number of 96,908,505 SNPs were called for 19 _Teosinte_ lines, of which 315,514 matched with GBS SNPs according to their v2 coordinates.

After checking `ref` and `alt` SNP calls, `301,249` sites are consistent between two datasets (concordant rate `r round(301249/315514, 3)*100`%). We then recoded the SNPs to `0, 1, 2` format, where the numbers indicated the copies of alternate alleles. 

The heterozygote error rate for GBS was estimated to be `49.1%`; and the homozygote error rate was estimated to be `1.7%`.

Finally, we calculate the missing rate and MAF for 19 teosintes of the two datasets, respectively. The results were stored in `cache/teo_gbs_wgs.RData`. In the following code chunk, we will plot the results.

```{r eval=T}
ob <- load("../cache/teo_gbs_wgs.RData")
par(mfrow=c(1,2)) 
hist(lmiss1, main="WGS (N=301,249)", ylim=c(0, 70000), col="#008080", 
     breaks=50, xlab="locus missing rate")
#abline(v=0.8, col="red", lty=2, lwd=3)
hist(lmiss2, main="GBS (N=301,249)", ylim=c(0, 70000), col="#008080", 
     breaks=50, xlab="locus missing rate")
#abline(v=0.8, col="red", lty=2, lwd=3)

par(mfrow=c(1,2)) 
hist(imiss1, main="WGS (N=19)",  col="#008080",  xlab="individual missing rate")
#abline(v=0.8, col="red", lty=2, lwd=3)
hist(imiss2, main="GBS (N=19)",  col="#008080",  xlab="individual missing rate")
#abline(v=0.8, col="red", lty=2, lwd=3)

par(mfrow=c(1,2)) 
hist(maf1, main="WGS (N=301,249)",  col="#008080",  xlab="MAF")
#abline(v=0.8, col="red", lty=2, lwd=3)
hist(maf2, main="GBS (N=301,249)",  col="#008080",  xlab="MAF")
#abline(v=0.8, col="red", lty=2, lwd=3)

```





# Comparing GBS vs. WGS




----------

